Automattic/dserve
# Adds support to run arbitrary docker images

## Background

In order to run tests in TeamCity, we need to change how `calypso.live` works. Instead of asking `dserve` to build an image based on a commit hash, we will build the image somewhere else, push it to the docker repo and ask `dserve` to download it.

See p3topS-Ok-p2 for more info

## Changes

This PR adds support for downloading and running arbitrary images. It adds two new URLs:

* `http://calypso.live?image=<image-name>`. Allows the client to specify an image to be downloaded. It can be an image from any docker registry that doesn't require credentials (limited to the registries defined in the config). It will download said image, create a container for it, and redirect the user to the URL below.
* `http://container-<container>.calypso.live`. Starts the container `<container>` as needed, and proxies all requests to it.

## Testing

- Checkout this branch and start `dserve` (it may take a while to update `wp-calypso` repo).
- Go to the URL specified in 270c2-pb. It should download it and redirect to Calypso.

Try a few variants:

- Load an image that doesn't exist locally
- In a separate tab, load an image that is being downloaded.
- Go to an image that exists locally but there are no containers for it
- Go to a container that exists but it is not started
- Go to a started container

Useful docker commands to test those variants:

- List downloaded images: `docker image ls`.
- Delete an image: `docker image rm <id> && docker image prune`
- List all containers `docker ps -a`
- Stop a running container `docker stop <id>`
- Delete a container `docker rm <id>`

## Pending things

I wanted to raise this PR when the core of the solution is done to get earlier feedback. There are still a few pending things (some of them can be addressed after this PR is merged):
- [x] touch containers so the process that clean up unused containers works as expected
- [ ] ensure that all errors are correctly propagated through the middleware layers
- [x] implement a query parameter to delete existing containers in case something breaks
- [ ] think about what info we want to add (if any) to the /debug dashboard or other debug URLs
